lib/IDE/Utils.cpp

```diff
@@ -1252,12 +1252,55 @@ Expr *swift::ide::getBase(ArrayRef<Expr *> ExprStack) {
   return Base;
 }
 
-bool swift::ide::isDynamicCall(Expr *Base, ValueDecl *D) {
-  auto TyD = D->getDeclContext()->getSelfNominalTypeDecl();
-  if (!TyD)
+static bool isDeclSyntacticallyOverridable(ValueDecl *D) {
+  auto *NTD = D->getDeclContext()->getSelfNominalTypeDecl();
+  if (!NTD)
     return false;
 
-  if (isa<StructDecl>(TyD) || isa<EnumDecl>(TyD) || D->isFinal())
+  if (!(isa<ClassDecl>(NTD) || isa<ProtocolDecl>(NTD)) || D->isFinal() ||
+      NTD->isFinal() || NTD->isActor())
+    return false;
+
+  if (auto *Accessor = dyn_cast<AccessorDecl>(D)) {
+    switch (Accessor->getAccessorKind()) {
+    case AccessorKind::Get:
+    case AccessorKind::Set:
+      break;
+    case AccessorKind::WillSet:
+    case AccessorKind::DidSet:
+    case AccessorKind::Address:
+    case AccessorKind::MutableAddress:
+    case AccessorKind::Read:
+    case AccessorKind::Modify:
+      return false;
+    }
+  }
+
+  return true;
+
+}
+bool swift::ide::isDeclOverridable(ValueDecl *D) {
+  if (!isDeclSyntacticallyOverridable(D))
+    return false;
+
+  // Specifically checking the decl here, so include a static check. Note that
+  // this is not valid when looking at a reference, since the base may be to a
+  // `MetaType`.
+  if (auto *FD = dyn_cast<FuncDecl>(D)) {
+    if (FD->isStatic() &&
+        FD->getCorrectStaticSpelling() == StaticSpellingKind::KeywordStatic)
+      return false;
+  } else if (auto *ASD = dyn_cast<AbstractStorageDecl>(D)) {
+    if (ASD->isStatic() &&
+        ASD->getCorrectStaticSpelling() == StaticSpellingKind::KeywordStatic)
+      return false;
+  }
+
+  return true;
+}
+
+bool swift::ide::isDynamicRef(Expr *Base, ValueDecl *D) {
+  if (!isDeclSyntacticallyOverridable(D))
```

Comment:
Any reason why this is not `isDeclOverridable`?

Reply:
Should I move the comment in `isDeclOverridable` to here? ie.
```
  // Specifically checking the decl here, so include a static check. Note that
  // this is not valid when looking at a reference, since the base may be to a
  // `MetaType`.
```
There's also a comment on `isDynamicRef` itself that explains it more.

Reply:
That one is:
```
/// A simple example is `SomeType.classMethod()`. `classMethod`
/// is itself overridable, but that particular reference to it *has* to be the
/// one in `SomeType`. Contrast that to `type(of: foo).classMethod()` where
/// `classMethod` could be any `classMethod` up or down the hierarchy from the
/// type of the \p Base expression.
```